### PR TITLE
[MM-15639] Add config setting to explicitly define which IP headers are trusted

### PR DESCRIPTION
**Pull request by DSchalla**
#### Summary
Currently we always trust the `X-Forwarded-By` and `X-Real-Ip` header as an authority what the clients IP address is. The problem with this behavior is when Mattermost is running without a proxy, a client can essentially send those headers himself and bypass rate limiting and/or the audit log. For environments that use a reverse proxy this problem does not exist, if the headers are set by NGINX itself. However, if `X-Real-Ip` is used and `X-Forwarded-By` is not filtered by the proxy it will essentially be overwritten.

This pull request adds a new configuration called `TrustedProxyIPHeader` that is a string array, allowing the administrator to pass an array of headers that will be checked one by one for IP addresses (order is important). All other headers are ignored.

New configs will have this set by default to `[]`, meaning that no header will be trusted. Current installations without the config entry will have it set to `X-Forwarded-By, X-Real-Ip`, copying the current behavior.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-15639

### CODE REVIEW

File: app/ratelimit.go [hunk 1]
```diff
@@ -23,9 +23,10 @@ type RateLimiter struct {
 	useAuth              bool
 	useIP                bool
 	header               string
+	trustedProxyIPHeader []string
 }
 
-func NewRateLimiter(settings *model.RateLimitSettings) (*RateLimiter, error) {
+func NewRateLimiter(settings *model.RateLimitSettings, trustedProxyIPHeader []string) (*RateLimiter, error) {
```
Comment (lieut-data) [inline app/ratelimit.go]: It doesn't look like we actually save `trustedProxyIPHeader` to the constructed `RateLimiter`?

File: app/ratelimit_test.go [hunk 2]
```diff
@@ -27,15 +27,15 @@ func genRateLimitSettings(useAuth, useIP bool, header string) *model.RateLimitSe
 
 func TestNewRateLimiterSuccess(t *testing.T) {
 	settings := genRateLimitSettings(false, false, "")
-	rateLimiter, err := NewRateLimiter(settings)
+	rateLimiter, err := NewRateLimiter(settings, nil)
```
Comment (lieut-data) [inline app/ratelimit_test.go]: Need a test for when this value is `non-nil`.

File: model/config.go [hunk 3]
```diff
@@ -434,6 +435,10 @@ func (s *ServiceSettings) SetDefaults() {
 		s.Forward80To443 = NewBool(false)
 	}
 
+	if s.TrustedProxyIPHeader == nil {
```
Comment (lieut-data) [inline model/config.go]: We're close to merging https://github.com/mattermost/mattermost-server/pull/10551 -- could we get the version in `master` to adopt the semantics of that PR for deciding which way to flip the defaults? (I realize we'll need to re-create your current code for the cherry-pick).

File: utils/utils.go [hunk 4]
```diff
@@ -68,19 +66,21 @@ func StringSliceDiff(a, b []string) []string {
...
-func GetIpAddress(r *http.Request) string {
+func GetIpAddress(r *http.Request, trustedProxyIPHeader []string) string {
 	address := ""
 
-	header := r.Header.Get(model.HEADER_FORWARDED)
-	if len(header) > 0 {
-		addresses := strings.Fields(header)
-		if len(addresses) > 0 {
-			address = strings.TrimRight(addresses[0], ",")
+	for _, proxyHeader := range trustedProxyIPHeader {
```
Comment (lieut-data) [inline utils/utils.go]: Looks like `HEADER_REAL_IP` and `HEADER_FORWARDED` have different encoding semantics, based on the old parsing behaviour. I /think/ the `Fields` and `TrimRight` amount to the same thing, but I wonder if we want to be explicit and handle each type instead? And do we really want to try to parse other headers which potentially different encoding semantics?

File: utils/utils.go [hunk 5]
```diff
@@ -68,19 +66,21 @@ func StringSliceDiff(a, b []string) []string {
...
+			if len(addresses) > 0 {
+				address = strings.TrimRight(addresses[0], ",")
+			}
 		}
-	}
 
-	if len(address) == 0 {
-		address = r.Header.Get(model.HEADER_REAL_IP)
+		if len(address) > 0 {
+			break
```
Comment (lieut-data) [inline utils/utils.go]: Would it be cleaner to just `return address` here?

File: utils/utils_test.go [hunk 6]
```diff
@@ -98,12 +98,34 @@ func TestGetIpAddress(t *testing.T) {
...
 		RemoteAddr: "10.2.0.1:12345",
 	}
 
-	assert.Equal(t, "10.2.0.1", GetIpAddress(&httpRequest5))
+	assert.Equal(t, "10.2.0.1", GetIpAddress(&httpRequest5, []string{"X-Forwarded-For", "X-Real-Ip"}))
+
+	// Test with both headers, but both untrusted
+	httpRequest6 := http.Request{
+		Header: http.Header{
+			"X-Forwarded-For": []string{"10.3.0.1"},
```
Comment (lieut-data) [inline utils/utils_test.go]: Can we add some coverage for the comma-encoded versions of this header?

File: app/ratelimit.go (hunk 1 above)
Comment (DSchalla) [reply app/ratelimit.go]: Good catch, thanks!

File: model/config.go (hunk 3 above)
Comment (DSchalla) [reply model/config.go]: This will need to go into 5.12, so ideally it should be merged today/tomorrow. Is there any timeline for #10551?

File: utils/utils.go (hunk 5 above)
Comment (DSchalla) [reply utils/utils.go]: Absolutely!

File: utils/utils_test.go (hunk 6 above)
Comment (DSchalla) [reply utils/utils_test.go]: It's done in https://github.com/mattermost/mattermost-server/pull/10907/files/0f2d1d71c4c50fe3b53d63005fcb9a436594c225#diff-8d029438ee1ad6087ae32de5b0636ccbR74 - Is there something you'd like to see additionally?

File: utils/utils.go (hunk 4 above)
Comment (DSchalla) [reply utils/utils.go]: I've considered it as well, but the problem is that `X-Real-Ip` is not standardized and I couldn't find a lot of information about it. I intentionally decided against implementing the special case, since the current code works fine with `X-Real-Ip` as well:
https://play.golang.org/p/eF0H4uLEsE1

Since it's not standardized some people might use it in a similar fashion, where this approach would still work, while the previous one would break.

File: app/ratelimit_test.go (hunk 2 above)
Comment (DSchalla) [reply app/ratelimit_test.go]: I've added some unit tests to cover all cases, also some on the key generation itself.

**Comment by DSchalla** [comment]
Thanks for the in-depth review @lieut-data! Yes, I absolutely agree that this is also problematic and one of the reasons why a install wizard would be beneficial, in the end both behaviors are problematic and there is no bullet proof solution but either helping the user through an interactive wizard or proper documentation sadly. Happy about any proposals!

Some sort of system notifications for admins after detecting a reverse proxy and check if the headers aren't configured would be useful I guess.

File: utils/utils_test.go (hunk 6 above)
Comment (DSchalla) [reply utils/utils_test.go]: Sorry, understood that you've meant for the multiple version, but untrusted and trusted. Added in the newest commit.

File: app/app.go [hunk 7]
```diff
@@ -134,7 +134,8 @@ func (a *App) HTMLTemplates() *template.Template {
 }
 
 func (a *App) Handle404(w http.ResponseWriter, r *http.Request) {
-	mlog.Debug(fmt.Sprintf("%v: code=404 ip=%v", r.URL.Path, utils.GetIpAddress(r)))
+	ipAddress := utils.GetIpAddress(r, a.Config().ServiceSettings.TrustedProxyIPHeader)
+	mlog.Debug(fmt.Sprintf("%v: code=404 ip=%v", r.URL.Path, ipAddress))
```
Comment (crspeller) [inline app/app.go]: Can we fix this to use structured logging?

File: web/web.go [hunk 8]
```diff
@@ -62,7 +62,7 @@ func CheckClientCompatability(agentString string) bool {
 func Handle404(config configservice.ConfigService, w http.ResponseWriter, r *http.Request) {
 	err := model.NewAppError("Handle404", "api.context.404.app_error", nil, "", http.StatusNotFound)
 
-	mlog.Debug(fmt.Sprintf("%v: code=404 ip=%v", r.URL.Path, utils.GetIpAddress(r)))
+	mlog.Debug(fmt.Sprintf("%v: code=404 ip=%v", r.URL.Path, utils.GetIpAddress(r, config.Config().ServiceSettings.TrustedProxyIPHeader)))
```
Comment (crspeller) [inline web/web.go]: Structured logging here too.

File: model/config.go (hunk 3 above)
Comment (crspeller) [reply model/config.go]: The problem here is that without the defaults.json this will be the default for new installs. Is this a problem solved by https://github.com/mattermost/mattermost-server/pull/10551?

Comment (lieut-data) [reply model/config.go]: Go ahead and merge -- I'll work with @reflog to iterate on the changes in the linked PR instead.

Comment (lieut-data) [reply model/config.go]: /Just/ missed your comment, @crspeller -- the changes in https://github.com/mattermost/mattermost-server/pull/10551 should be fungible with respect to our current use of `default.json`. The code would look something like:

```go
if isUpdate {
    s.TrustedProxyIPHeader = []string{HEADER_FORWARDED, HEADER_REAL_IP}
} else {
    s.TrustedProxyIPHeader = ""
}
```

where `isUpdate` is `true` whenever `ServiceSettings.SiteURL` is non-nil.

File: app/app.go (hunk 7 above)
Comment (DSchalla) [reply app/app.go]: Yeah, good call. Should we potentially open a campaign to fix those? I don't know.
